services/api/service.go

```diff
@@ -1733,6 +1739,45 @@ func (api *RelayAPI) checkSubmissionSlotDetails(w http.ResponseWriter, log *logr
 		return false
 	}
 
+	// Do a check in memchache for the slot being with a registered provider. If miss, go to redis. if miss again go to postgres.
+
+	// Find the duty for the submission slot
+	api.proposerDutiesLock.RLock()
+	duty := api.proposerDutiesMap[submission.BidTrace.Slot]
+	api.proposerDutiesLock.RUnlock()
+	if duty == nil {
+		log.Info("no duty found for submission slot")
+		api.RespondError(w, http.StatusBadRequest, "no duty found for submission slot")
+		return false
+	}
+
+	if api.opts.MevCommitFiltering {
+		// Check if validator is registered
+		start := time.Now()
+		isValidatorRegistered, err := api.datastore.IsMevCommitValidatorRegistered(common.NewPubkeyHex(duty.Entry.Message.Pubkey.String()))
+		if err != nil {
+			log.WithError(err).Error("Failed to check validator registration")
+			api.RespondError(w, http.StatusInternalServerError, "Internal server error")
+			return false
+		}
+
+		if isValidatorRegistered {
+			isBuilderRegistered, err := api.datastore.IsMevCommitBlockBuilder(common.NewPubkeyHex(submission.BidTrace.BuilderPubkey.String()))
+			if err != nil {
+				log.WithError(err).Error("Failed to check builder registration")
+				api.RespondError(w, http.StatusInternalServerError, "Internal server error")
+				return false
+			}
+			if !isBuilderRegistered {
+				// TODO: Implement caching strategy for builder registration status
```

Comment:
Instead you can listen to the registration events and store it in a cache to check against. I think your entire MevCommitClient can just listen for events and store things locally to return. That way it will be way more efficient. Similar to what we do in mev-commit.

Reply:
Yeah, I've set the approach to now subscribe to events and when a builder is registered, we add it into the redis cached hashmap. We periodically refresh and remove builder registrations that are no longer valid.